### ISSUE TITLE
`magit-auto-revert-mode` should be off by default!!

### ISSUE THREAD

**Issue by elibarzilay** [magit/magit#1783]
(Double bangs in the subject intended.)

I'm surprised that even though there is now a (very welcome) option to turn auto-revert mode off, it is on by default.  IMO, it is an extremely bad idea to do this (or to even have this option in the first place).

The reason is that it can lead to a disaster of losing your work: save file, do some git stuff on it via magit (like commit then reverse) and the file is refreshed behind your back:

* Without asking you (eg, when a saved file is touched on disk, and you try to edit it, Emacs asks you about reverting the file).

* Without any way of recovering the lost information (eg, I change the above Emacs feature and make it always auto-revert when you touch the unsaved buffer -- **but** I do this as a change that is pushed to the undo list, which makes the previous content + edits easily available).

* Without any indication!  This looks like a feature that was carefully done: when a buffer is reverted, it is done in a way that leaves the cursor in the same place, and even an active region stays the same -- but whatever part of your buffer that has just been obliterated is outside of the visible part, you won't even see it.  Not to mention that magit will happily inflict such damage on any unfortunate buffers making it even easier to miss such a disaster.

The usual caveats apply about the lost information being kept in a volatile running process, about information kept by git itself, even the fact that **actually** using the undo information is a semi-lost art, etc etc.  I have complained about this in the past on the mailing list, describing one case where I lost a bunch of stuff as a result -- but there are many ways to get such lossage.

**Comment by tarsius** [comment]
I have just released `1.4.0` it now nags every user at startup, so that everyone makes a conscious decision to use this feature or not. 

I think you overestimate the risk of dataloss. It's there but it is not all that likely that this happens.

> Without asking you (eg, when a saved file is touched on disk, and you try to edit it, Emacs asks you about reverting the file).

We cannot do that because when "Magit changes a file" it does so by calling Git not by using functions built into Emacs. So every change Magit makes is seen as a change made outside of Emacs.

> Without any way of recovering the lost information [...]

I don't quite understand what you are saying in the part I have elided, but the part I quoted is not true - it can be undone using the `undo` command, as you mention later too.

> Without any indication!

This is the reason why users are now asked whether they want to keep automatic reverts on.

> even the fact that actually using the undo information is a semi-lost art

I strongly doubt that.

----

Users are now asked whether they want to turn this off. I strongly recommend they don't, it would lead to annoying behavior. For example a checkout or a hard reset would not be fully reflected in Magit/Emacs. And for users who do this sort of thing a lot, Emacs would start to constantly ask about outdated buffers.

One should not e.g. do a hard reset relying on the previous information remaining in a file-visiting buffer and then saving the buffer again, to keep it around and further work with it. Emacs would ask the users questions. It would be much better to do a soft reset.

Note that the user only risks losing data when doing such a "through this away" action like reset. If she attempted to e.g. checkout another branch, then Git itself would refuse to do it. The file would not change on disk and Emacs/Magit would not revert the buffer which is visiting it.

**Comment by elibarzilay** [comment]
> I have just released 1.4.0 it now nags every user at startup, so that
> everyone makes a conscious decision to use this feature or not.

Thanks!  -- I hope that the question is clear enough, but this is much
better than silently using that as a default.


> I think you overestimate the risk of dataloss. It's there but it is
> not all that likely that this happens.

I don't see why you'd say that it's an overestimate -- in my case, I had
minor bits lost over a few months before I tracked this down to magit
auto-refreshing buffers without telling me.  (A behavior that I would
never expect, which is why it took me so much time to realize that it's
actually happenning, and track it down.)  The later discussion (when
global auto-revert mode was turned on by mistake) was also described as
something that took a while to track down.  So there are at least two
documented cases -- and, at least in my case, the way in which
information was lost, was very easy to get to.


> > Without asking you (eg, when a saved file is touched on disk, and you
> > try to edit it, Emacs asks you about reverting the file).
> 
> We cannot do that because when "Magit changes a file" it does so by
> calling Git not by using functions built into Emacs. So every change
> Magit makes is seen as a change made outside of Emacs.

I'm not saying that magit should ask you for permission to change a file
on disk: that's obviously the whole point of git and therefore magit.
What I'm saying is that changing the content of a *buffer* is not
something that should be done without user confirmation.  That's what
Emacs does when the file you're working on had changed on disk: it tells
you that he file had been modified and asks you if you want to revert to
the changed version.


> > Without any way of recovering the lost information [...]
> 
> I don't quite understand what you are saying in the part I have
> elided, but the part I quoted is not true - it can be undone using the
> undo command, as you mention later too.

No it cannot!!  When a file gets (auto) reverted the buffer is
essentially re-initialized and re-loaded, which means that the undo
history from before the revert is gone.  This is one major reason why
Emacs asks before doing that.  As I mentioned, I hacked that to revert
automatically, but I make it beep with a message and also I make the
revert be an undoable operation with the previous history still
accessible.  The principle of not losing past edits holds in both cases,
and fails in magit.  (IIUC, it also fails with the Emacs auto-revert
mode, but I wouldn't dream about even trying it.)

> > even the fact that actually using the undo information is a
> > semi-lost art
> 
> I strongly doubt that.

(In many cases I've seen people complaining about undo in Emacs getting
the undo operations onto the undo list.  In all of these people would
conclude that Emacs is horrible exactly because it doesn't do the
castrated kind of linear undo line that you get in popular editors.)

> Users are now asked whether they want to turn this off. I strongly
> recommend they don't, it would lead to annoying behavior. For example
> a checkout or a hard reset would not be fully reflected in
> Magit/Emacs.

That's the way it works with many, many other features that deal with
external updates to files that are open in Emacs...  (And obviously I'd
strongly recommend not turning it on, it's obviously to a universally
liked feature.)


> And for users who do this sort of thing a lot, Emacs would start to
> constantly ask about outdated buffers.

Only once for each file, on a first edit after an on-disk change...
That's not what I'd classif as "constantly ask".


> One should not e.g. do a hard reset relying on the previous
> information remaining in a file-visiting buffer and then saving the
> buffer again, to keep it around and further work with it. Emacs would
> ask the user many questions. It would be much better to do a soft
> reset.

I commonly do a reset to drop some quick half-done edits and be able to
easily catch up with upstream changes.  Later on when I go back to that
file I still have the aborted state in Emacs, which allows me to easily
continue where I left off.  This is in sharp contrast to doing a soft
reset, then do a pull, then realize that I have changes and stash them,
re-pull, pop the stash and deal with possible conflicts (extremely
distracting) -- overall a much lengthier process.

That's not to say that I never use a soft/mixed reset and stash -- it's
very useful for longer term changes or if I have more changes than can
be stored in short-term-memory, but for quick edit-sync-continue-editing
jobs my way is *much* faster.


> Note that the user only risks losing data when doing such a "through
> this away" action like reset. If she attempted to e.g. checkout
> another branch, then Git itself would refuse to do it. The file would
> not change on disk and Emacs/Magit would not revert the buffer which
> is visiting it.

Not really -- this applies to any commands that can change files, where
I have such a file open, which is much more common than just resetting.
Doing a checkout, or any partial reversal of some changes, or even a
simple pull can end up with new contents that will overwrite what I have
in the buffer: since I may have a file opened and *expecting* it to
remain the same after the pull -- intentionally unsynced with the file
contents.

As an example for a workflow that I use often: I have some files that
I'm working on open, and if a pull updated them I can now easily grab
the buffer contents before it's reverted for an easy access to the
pre-pull contents.  In this specific case I can get the same from the
git history, but that's again more work than just having the previous
contents sitting in the buffer waiting for me to decide what to do with
it.
